### PR TITLE
Harness implemented as global mocha test.

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @AlexanderMoskovkin
Take a look, please (not for merging)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -63,8 +65,11 @@
     "gulp-uglify": "^1.5.1",
     "gulp-util": "^3.0.7",
     "gulp-webmake": "0.0.4",
+    "http": "0.0.0",
+    "path": "^0.12.7",
```
Comment (AlexanderMoskovkin) [inline package.json]: delete 'path'

File: package.json [hunk 2]
```diff
@@ -63,8 +65,11 @@
     "gulp-uglify": "^1.5.1",
     "gulp-util": "^3.0.7",
     "gulp-webmake": "0.0.4",
+    "http": "0.0.0",
```
Comment (AlexanderMoskovkin) [inline package.json]: delete it

File: package.json [hunk 3]
```diff
@@ -63,8 +65,11 @@
...
+    "path": "^0.12.7",
     "publish-please": "^1.0.1",
     "request": "^2.58.0",
+    "saucelabs-connector": "0.0.1",
```
Comment (AlexanderMoskovkin) [inline package.json]: ^0.0.1

File: test/functional/config.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  "browsers": [
```
Comment (AlexanderMoskovkin) [inline test/functional/config.json]: chrome, ie11, ff

check ie args

File: test/functional/fixtures/before-test.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+var slConnector = null;
+var slBrowsers  = null;
+
+var configPath = '../config.json';
+
+before(function () {
+    var config = JSON.parse(read(configPath));
+
+    return createTestCafe(config.testCafe.hostname, config.testCafe.port1, config.testCafe.port2)
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: var hostname = process.env.TRAVIS ? `http://${os.hostname()}:` : config.testCafe.hostname;

File: test/functional/fixtures/before-test.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+var getReportErrorByBrowsers = require('../helpers/report-filter.js');
+
+var testCafe     = null;
+var browsersInfo = null;
+
+var slConnector = null;
+var slBrowsers  = null;
+
+var configPath = '../config.json';
+
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: SAUCELABS_SETTINGS

File: test/functional/fixtures/before-test.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                        });
+                });
+
+                startPromise = Promise.all(openBrowserPromises);
+            }
+
+            return startPromise;
+        })
+        .then(function () {
+            global.runTests = function (fixture, testName) {
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: add the `should fail` argument

**Comment by georgiy-abbasov** [comment]
Error assertion:

Default mocha spec report
Default mocha spec reporter

Our simplified report inherited from mocha's spec.

Simplified report

File: test/functional/fixtures/before-test.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+var getErrorByBrowsersInfo = require('../helpers/report-helper.js');
+var config                 = require('../config.js');
+
+var testCafe     = null;
+var browsersInfo = null;
+
+var slConnector = null;
+var slBrowsers  = null;
+
+var SAUCELABS_SETTINGS = {
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: move this to `config.js` too

File: test/functional/config.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        },
+        {
+            platform:    'Windows 10',
+            browserName: 'chrome',
+            alias:       'chrome'
+        }
+    ],
+    testCafe: {
+        hostname: '127.0.0.1',
+        port1:    1335,
```
Comment (AlexanderMoskovkin) [inline test/functional/config.js]: Not all of ports are available on some of saucelabs wms (for example on ios). 
Let's use `2000, 2001, 3000, 3001` ports here.

File: test/functional/helpers/report-helper.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                        })
+                        .join(' ');
+                }
+            }
+        });
+
+    return testError;
+}
+
+module.exports = getErrorByBrowsersInfo;
```
Comment (AlexanderMoskovkin) [inline test/functional/helpers/report-helper.js]: Is this name confirmed by a techwriter?

File: test/functional/helpers/report-helper.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
+function checkTestRelevanceByBrowserAlias (test, browserAlias) {
```
Comment (AlexanderMoskovkin) [inline test/functional/helpers/report-helper.js]: The module exports just one function. So, it should be called by this function name. (For example, if the module exports just a function that reads any file, the module should be called `read-file.js`). Discuss with a tech writer please

File: test/functional/fixtures/before-test.js [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+            }
+
+            return startPromise;
+        })
+        .then(function () {
+            global.runTests = function (fixture, testName, opts) {
+                var report      = '';
+                var runner = testCafe.createRunner();
+                var fixturePath = path.join(path.dirname(caller()), fixture);
+
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: define the `shouldFail` var here to define what opts we expect

``` js
var shouldFail = opts && opts.shouldFail;
```

File: test/functional/fixtures/before-test.js [hunk 13]
```diff
@@ -0,0 +1,146 @@
+var createTestCafe         = require('../../../lib');
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: require our internal modules after the outside modules

File: test/functional/fixtures/before-test.js [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+var SAUCELABS_SETTINGS = {
+    username:  process.env.SAUCE_USERNAME,
+    accessKey: process.env.SAUCE_ACCESS_KEY,
+    build:     process.env.TRAVIS_JOB_ID || '',
+    tags:      [process.env.TRAVIS_BRANCH || 'master'],
+    name:      'testcafe functional tests'
+};
+
+before(function () {
+    var hostname = process.env.TRAVIS ? os.hostname() : config.testCafe.hostname;
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: move this definition to `config.js` too

File: test/functional/fixtures/before-test.js (hunk 13 above)
Comment (georgiy-abbasov) [inline test/functional/fixtures/before-test.js]: Should be there a newline between them (internal and outside modules) ?

Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: no

File: test/functional/fixtures/before-test.js [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    return createTestCafe(hostname, config.testCafe.port1, config.testCafe.port2)
+        .then(function (tc) {
+            var startPromise        = null;
+            var openBrowserPromises = null;
+
+            testCafe = tc;
+
+            browsersInfo = config.browsers
+                .filter(function (browser) {
+                    return SAUCELABS_SETTINGS || !browser.remoteOnly;
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]:
``` js
return SAUCELABS_SETTINGS || ...
```

always returns `true`. You defined this above. What does it mean here?

File: test/functional/fixtures/before-test.js [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+                    return browserNatives.getBrowserInfo(browserInfo.settings.alias)
+                        .then(function (browser) {
+                            return browserNatives.open(browser, browserInfo.connection.url);
+                        });
+                });
+
+                startPromise = Promise.all(openBrowserPromises);
+            }
+
+            return startPromise;
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: it's not understandable 'start what?' from this var name.
I'd prefer to get rid of it:

``` js
if(process.env.TRAVIS) {
    ...
    return slConnector.connect().then()....
}

openBrowserPromises = ...
return Promise.all(openBrowserPromises);
```

File: test/functional/fixtures/before-test.js (hunk 15 above)
Comment (georgiy-abbasov) [inline test/functional/fixtures/before-test.js]: I've tried this way, which you recommend, before, but eslint showed error Unexpected 'else' after 'return'.

File: test/functional/fixtures/before-test.js (hunk 16 above)
Comment (georgiy-abbasov) [inline test/functional/fixtures/before-test.js]: I've tried this way, which you recommend, before, but lint showed error " Unexpected 'else' after 'return' "

Comment (AlexanderMoskovkin) [inline test/functional/fixtures/before-test.js]: sure,

``` js
if(a)
    return 1;
else    //eslint error
    return 2;
```

``` js
if(a)
    return 1;

return 2; // it's all ok
```

Comment (georgiy-abbasov) [inline test/functional/fixtures/before-test.js]: thanks!

File: test/functional/helpers/report-helper.js [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    return only && !skip;
+}
+
+function filterErrorsByUserAgents (errors, userAgents) {
+    return errors.filter(function (error) {
+        return userAgents.indexOf(error.split('\n')[0]) > -1;
+    });
+}
+
+function getErrorByBrowsersInfo (report, browsersInfo) {
```
Comment (AlexanderMoskovkin) [inline test/functional/helpers/report-helper.js]: function getTestError (testReport, browsersInfo)
//returns a string with error

File: test/functional/helpers/report-helper.js (hunk 11 above)
Comment (georgiy-abbasov) [inline test/functional/helpers/report-helper.js]: \cc @VasilyStrelyaev, take a look at report-helper.js, please.

Comment (VasilyStrelyaev) [inline test/functional/helpers/report-helper.js]: You are filtering browsers not tests, so maybe they are what may be relevant or irrelevant?
`checkTestRelevanceByBrowserAlias` -> `isBrowserRelevant`

Comment (VasilyStrelyaev) [inline test/functional/helpers/report-helper.js]: `filterErrorsByUserAgents` is a bit simpler than everything else here, so we can go with it (unless `filterErrors` is enough)

Comment (VasilyStrelyaev) [inline test/functional/helpers/report-helper.js]: report-helper.js
/r-

Comment (AlexanderMoskovkin) [inline test/functional/helpers/report-helper.js]: I vote for `filterError`. We can understand _filter by what_ from the function arguments

Comment (VasilyStrelyaev) [inline test/functional/helpers/report-helper.js]: yep!

**Comment by georgiy-abbasov** [comment]
fpr

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
thanks

**Pull request closed without merging**